File: tracking/matlab/locateMid.m
%% Finds a place in the middle of the shaft where the gradient changes from
%% dark to light. 
% given: ex_loc, ex_angle, ex_rho
% ex_loc should be the point we expect to return
% ex_angle is the initial estimate for the angle of the sleeves/shaft with
%  the y-axis
% ex_rho is the estimate of the perpendicular distance from the midline to
%  the image origin (see Wikipedia page for Hough transform)

%% initial vars
pre = 'knottie';
nx = 1280;
ny = 960;
maskSide = 350;

% this is the expected grayscale value of the black shaft, after the image
% is smoothed. For some reason convolution scales the entire image down.
ex_gray = .1;           
% how much variation to allow before we rule out a color as belonging to
% the shaft
gray_tol = .15;

% We'll allow lines that are within this much of ex_rho
rho_tol = 200;
% Allow lines within this many degrees of ex_angle.
angle_tol = 20;
% expected width of shaft in image. 
armw_min = 60;
armw_max = 80;
% when finding parallel lines, allow to differ in angle by these many
% degrees.
angd_tol = 1;

% base tracking directory
directory = RemoteHomeDirectory;

% where the images are kept
imDir = 'data/2010Jul23/gray';
filename = [pre int2str(k) '-' int2str(imNum) '.tif']
im = imread([directory '/' imDir '/' filename]);

% grayscale image if not already 
if length(size(im)) == 3
    img = rgb2gray(im);
else
    img = im;
end

cannythresh = [.05, .1];

[bw, th] = edge(img, 'canny', cannythresh, 2);
figure; imshow(bw);

%% Do Hough transform and choose the most probable lines
%% allpeaks is an nx2 array, wherein thetas(allpeaks(i,2)) is the angle of
%% line i and rhos(allpeaks(i,1)) is the perpendicular distance to the
%% origin of line i. 

% angle discretization
T = -90:.1:89;
% do Hough transform
[H,thetas,rhos] = hough(bw, 'RhoResolution', .5, 'Theta', T);

% thresh = 0.20*max(H(:));
% how many votes a line needs to have in order to be considered 
thresh = 55;
% suppression neighborhood: if a line is taken to be a peak, ignore 
nhood = [13,19];
npeaks = 20;
minlen = 110;
% find lines that have sufficiently many votes.
allpeaks = houghpeaks(H, npeaks, 'Threshold', thresh, 'nhoodsize', nhood);

P = allpeaks;
T = thetas;
figure; imshow(bw, opt, 50); hold on;
showhoughpeaks

%% filter peaks
%% Choose lines that are close to the expected angle, and whose
%% perpendicular distances are close to expected.
%% "pekes" contains the filtered peaks.

bool = angleDiffBlind(ex_angle, thetas(allpeaks(:,2)), 'deg') < angle_tol ...
    & abs(rhos(allpeaks(:,1)) - ex_rho) < rho_tol;

pekes = allpeaks(bool, :);
P = pekes;
figure; imshow(bw, opt, 50); hold on;
showhoughpeaks

%% find parallel lines the right width apart. 
%% pekes1(i,:) and pekes2(i,:) should parametrize a pair of parallel lines

Ti = pekes(:,2);
angdMatrix = zeros(length(Ti));
distMatrix = zeros(length(Ti));

for i=1:length(Ti)
    for j=(i+1):length(Ti)
        d = angleDiff(T(Ti(i)), T(Ti(j)), 'deg');
        angdMatrix(i,j) = d;
        angdMatrix(j,i) = d;
    end
end

rhoi = pekes(:,1);
for i=1:length(Ti)
    for j=(i+1):length(Ti)
        distMatrix(i,j) = abs(rhos(rhoi(i)) - rhos(rhoi(j)));
        distMatrix(j,i) = abs(rhos(rhoi(i)) - rhos(rhoi(j)));
    end
end

boolMatrix = (distMatrix < armw_max) & (distMatrix > armw_min) & (angdMatrix < angd_tol);
[ii, jj] = find(boolMatrix);
bb = ii < jj;
ii2 = ii(bb);
jj2 = jj(bb);
pekes1 = pekes(ii2,:);
pekes2 = pekes(jj2,:);

%% output
for i=1:length(ii2)
    P = [pekes1(i,:); pekes2(i,:)];
    figure; imshow(bw, opt, 50);
    showhoughpeaks
end

%% for each pair, find rho and theta of midline
xx = 1:nx;
rhomids = zeros(1,size(pekes1,1));
thmids = zeros(1,size(pekes1,1));
for i=1:size(pekes1, 1)
    pk1 = pekes1(i,:);
    pk2 = pekes2(i,:);
    r1 = rhos(pk1(1));
    r2 = rhos(pk2(1));
    rhomids(i) = .5*(r1+r2);    % average rhos
    thmids(i) = thetas(pk1(2)); % just take theta of line1; the idea is that the angles are so close this doesn't make a big difference
end

figure; imshow(bw); hold on;
for i=1:length(rhomids)
    m = -cot(thmids(i)*pi/180);
    b = rhomids(i)/sin(thmids(i)*pi/180);
    plot(1:nx, m*(1:nx) + b);
end

%% Put together of pool of candidate points. 
%% Candidate points must exceed the thresholds for directional similarity,
%% magnitude strength, and closeness of the color to the expected shaft
%% color. 
%%  Note: When calculating color similarity, use grayscale value of the point a
%%  few increments back on the midline, since the color before our chosen
%%  point must be the shaft color and the color after must be the band
%%  color.
%%
%%  Note 2: Grayscale values are taken from smoothed image, and so ex_gray
%%  should refer to the expected smoothed grayscale value. 

sig = 2; 
% note: the 1.5 multiplier is mostly for display purposes. The images tend
% to be dark.
imgsm = conv2(im2double(img)*1.5, fspecial('gaussian', [9,9],sig), 'same'); % smoothed image

dg = GaussianDerivative(sig);
dX = imfilter(im2double(img)*1.5, dg, sig, 'replicate', 'same'); 
dY = imfilter(im2double(img)*1.5, dg', sig, 'replicate', 'same'); 
dMags = sqrt(dX.*dX + dY.*dY);
imOris = atan2(dY, dX);
nconsider = 4;

dispim  = imgsm;

candidate_locs = [];
candMags = [];
candSims = [];
for i=1:length(rhomids)
    
    % get tX, the points on the midline. 
    s = sin(thmids(i)*pi/180);
    c = cos(thmids(i)*pi/180);
    R = [-s -c; c -s];
    
    dx = sin(thmids(i)*pi/180);
    sgn = dx/abs(dx);
    dy = cos(thmids(i)*pi/180);
    dx = sgn*dx;
    dy = -sgn*dy;
    
    nsteps = min(floor(nx/dx), floor(abs(ny/dy)));
    X = [2:nsteps-1; zeros(1,nsteps-2)];
    tX = R*X;
    
    icept = rhomids(i)/s;
    if icept >= 0
        tX = vadd(tX, [0; rhomids(i)/(s)]);
    else 
        tX = vadd(tX, [rhomids(i)/c;0]);
    end     
        
    % got the list of points to be looking at. Now compare their
    % values/gradients.
    inds = sub2ind(size(img), round(tX(2,:)), round(tX(1,:)));
%     dispim(inds) = 255;
    gradMags = dMags(inds);
    graddirs = [dX(inds) ;dY(inds)];
    graddirs = graddirs./([1;1]*normArray(graddirs));
    gradDirSims = (graddirs' * [dx; dy])'; 
    vals = imgsm(inds);
    
    % calculate magthresh to be a fraction of the max gradient (not
    % considering outliers);
    magOutlierThresh = mean(gradMags) + 2*std(gradMags);
    magthresh = .4*max(gradMags(gradMags < magOutlierThresh));
    gradMags(gradMags >= magOutlierThresh) = magOutlierThresh;

    simthresh = .9;
    figure; plot(gradMags); hold on
    hold on; plot(gradDirSims, 'g');
    plot([1, nsteps], magthresh*[   1 1], 'r');
    plot([1, nsteps], simthresh*[1 1], 'm');
    
    % figure out which points pass the thresholds, and take first
    % "nconsider." The reason for only considering the first few is that 
    % the later points tend to fall on the hand itself or some other 
    % object, and the algorithm has mistaken these for the right point. 
    % On the other hand, since the beginning of the midline tends to fall
    % in the middle of the sleeve, there aren't many strong gradients, and
    % so the first signals are the ones we want. 
    bool = gradMags > magthresh & gradDirSims > simthresh...
        & [true true true abs(vals(1:(end-3)) - ex_gray) < gray_tol];
    firstidx = find(bool, nconsider, 'first');
    
    candidate_locs = [candidate_locs, tX(:,firstidx)];
    candMags = [candMags gradMags(firstidx)];
    candSims = [candSims gradDirSims(firstidx)];
end

%% From the pool, choose one
%% Score by (dists-to-expected)/(gradient-magnitude). 
difs = vadd(candidate_locs, -ex_loc);
dists = normArray(difs);
[mn, mi] = min(dists./(candMags));

%% display
if true
    fig = figure; imshow(dispim, opt, 50); hold on;
    scatter(candidate_locs(1,:), candidate_locs(2,:), 'g');
    scatter(candidate_locs(1, mi), candidate_locs(2,mi), 'm');
end

% figure; imshow(dispim)

%% outputs
midpt = candidate_locs(:,mi);

pekes3 = unique([pekes1; pekes2], 'rows');
peak_thetas = [thetas(pekes3(:,2))];

% take care of some angle-wrapping business
peak_thetas(peak_thetas - ex_angle >= 90) = peak_thetas(peak_thetas - ex_angle >= 90) - 180;
peak_thetas(ex_angle - peak_thetas >= 90) = peak_thetas(ex_angle - peak_thetas >= 90) + 180;

avgAngle = mean(peak_thetas);
if avgAngle >= 90
    avgAngle = avgAngle - 180;
elseif avgAngle < -90
    avgAngle = avgAngle + 180;
end

avgRho = mean(rhomids);
thetaPeaks = [thetas(pekes1(:,2))' thetas(pekes2(:,2))'];
rhoPeaks= [rhos(pekes1(:,1))' rhos(pekes2(:,1))'];

%% other stuff
%% Prepare for next iteration.
ex_loc = midpt;
ex_angle = avgAngle;
ex_rho = avgRho;

File: tracking/cpp/locateMid.cpp
# include <iostream>
# include <opencv/cv.h>
# include <opencv/highgui.h>
# include "utils.h"

using namespace cv;

/*
 * Inputs: 
 *  filepath -- path to grayscale image
 *  ex_loc_x, ex_loc_y -- where you would expect to find the model origin in the image.
 *  ex_rho -- expected perpendicular distance to image origin. 
 *            Note: opencv's hough transform uses a different origin than Matlab's (negate Matlab rho?)
 *  ex_angle -- expected angle in degrees
 *  cannythresh_low, cannythresh_high -- parameters for the Canny edge detector.
 *  houghthresh -- how many votes a line needs to have in order to be considered 
 *  magthresh -- only points with gradient magnitude larger than this are put into the pool of candidates. 
 *  debug -- if this argument is present, then a lot of intermediate variables are saved to disk.
 * 
 * Example: 
 *  ./locateMid /home/jinna/surgical/tracking/data/2010Jul23/rect/knottie1-_rect573.ppm  507 459 100 -38
 */
int main(int argc, char** argv) {
    // inputs and parameters

    int numArgs = 10;
    if (argc < numArgs) {
        std::cerr << "Usage: exe filepath ex_loc_x ex_loc_y ex_rho ex_angle(deg) cannythresh_low cannythresh_high houghthresh magthresh [debug]" << std::endl;
        exit(1);
    }
    std::string filepath  = argv[1];
    Point2f ex_loc(atof(argv[2]), atof(argv[3]));
    float ex_rho = atof(argv[4]);
    float ex_angle = atof(argv[5]);
    double cannythresh[] = {atof(argv[6]), atof(argv[7])};
    int houghthresh = atoi(argv[8]);    
    float magthresh = atof(argv[9]);
    float simthresh = .9;
    bool left = true;
    bool debug = false;
    if (argc > numArgs) {
        std::string debugStr = argv[argc-1];
        if (debugStr == "d") {
            debug = true;
            std::cout << "debug mode" << std::endl;
        }
    }    
    
    Mat img;
    
    img = imread(filepath.c_str(), GRAYSCALE_IMAGE);
    if (img.data == NULL) {
        std::cerr << "there was an error reading the image" << std::endl;
        exit(1);
    }
    if (img.channels() != 1) {
        std::cerr << "img must be grayscale. channels= " << img.channels() << std::endl;
        exit(1);
    }
    
//     namedWindow("orig image", CV_WINDOW_AUTOSIZE);
//     imshow("orig image", img);
    
    int nx = img.size().width;
    int ny = img.size().height;
    float ex_gray = .1;
    float gray_tol = .15;
    float rho_tol = 50;
    float angle_tol = 2;
    float armw_min = 60;
    float armw_max = 80;
    float angd_tol = 1;
    float sig = 2;
//     float cannythresh[] = {.05, .1};
    
    // Smooth image
    Mat imgf(img.size(), CV_32FC1);
    img.convertTo(imgf, CV_32FC1);
    imgf = imgf/(255.0);
    
    Mat imgsm(img.size(), CV_32FC1);
    GaussianBlur(imgf, imgsm, Size(17,17), sig, sig);
//     namedWindow("blurred", CV_WINDOW_AUTOSIZE);
//     imshow("blurred", imgsm);
    if (debug) {
        saveToFile<float>(imgsm, "imgsm_locateMid.txt");
    }
    
    // Perform Canny detection
    // first make sure array is in right format
    Mat imgsm256(imgsm.size(), CV_8UC1);
//     imgsm = imgsm*255;    
    imgsm.convertTo(imgsm256, CV_8UC1, 255.0);
    Mat imgsm256_output(imgsm.size(), DataType<int>::type);
    imgsm256.convertTo(imgsm256_output, DataType<int>::type);
    if (debug) {
        saveToFile<int>(imgsm256_output, "imgsm256.txt");
    }
    
    Mat bw;
    std::cout << "cannythresh: " << cannythresh[0] << " " << cannythresh[1] << std::endl;
    Canny(imgsm256, bw, cannythresh[0], cannythresh[1]);
//     namedWindow("blurred2", CV_WINDOW_AUTOSIZE);
//     imshow("blurred2", imgsm256);
    
//     namedWindow("canny edges", CV_WINDOW_AUTOSIZE);
//     imshow("canny edges", bw);
    if (debug) {
        saveToFile<int>(bw, "bw.txt");
    }
    
    // Hough lines
    vector<Vec2f> lines;
    double rhoRes = .5;
    double thetaRes = .1;
    
    HoughLines(bw, lines, rhoRes, thetaRes*RADIANS_PER_DEGREE, houghthresh);
    std::cout << lines.size() << " lines found" << std::endl;
    
    // suppress neighborhods -- matlab does this automatically
    float rho_nhood = 5; // pixels
    float theta_nhood = 2*RADIANS_PER_DEGREE;
    bool yes[lines.size()];
    int total = 0;
    
    for( int i = 0; (unsigned)i < lines.size(); i++ )
    {
        float rho = lines[i][0];
        float theta = lines[i][1];
        
        bool take = true;
        for (int j=0; j<i; j++) {
            if (yes[j] && abs(rho - lines[j][0]) < rho_nhood && abs(theta - lines[j][1]) < theta_nhood) {
                take = false;
            }
        }
        yes[i] = take;
        if (take) {
            total++;
        }
    }
    
    vector<Vec2f> lines_sparse(total);
    int place = 0;
    for (int i=0; (unsigned)i<lines.size(); i++) {
        if (yes[i]) {
            lines_sparse[place][0] = lines[i][0];
            lines_sparse[place][1] = lines[i][1];
            place++;
        }
    }

    // display
//     Mat bw_disp2 = draw_lines(bw, lines);
//     namedWindow("all hough lines", CV_WINDOW_AUTOSIZE);
//     imshow("all hough lines", bw_disp2);

    
//     Mat bw_disp = draw_lines(bw, lines_sparse);
//     namedWindow("sparse hough lines", CV_WINDOW_AUTOSIZE);
//     imshow("sparse hough lines", bw_disp);

    // filter peaks according to closeness to expected angle and rho
    total = 0;
    bool yes_peaks[lines_sparse.size()];
    for (int i=0; (unsigned)i<lines_sparse.size(); i++) {
        float angle = lines_sparse[i][1]/((float)RADIANS_PER_DEGREE) ;
        float tmpdiff = abs(angle - ex_angle); // in degrees
        tmpdiff = fmod(tmpdiff, 180);
        float angleDiff = min(tmpdiff, 180-tmpdiff);
        float rhoDiff = abs(lines_sparse[i][0] - ex_rho) ;
        std::cout << "angle =" << angle << ", angleDiff=" << angleDiff << std::endl;
        std::cout << "rho=" << lines_sparse[i][0] << ", rhoDiff=" << rhoDiff << std::endl;
        if (angleDiff < angle_tol && rhoDiff < rho_tol) {
            yes_peaks[i] = true;   
            total++;
        } else {
            yes_peaks[i] = false;
        }   
    }
    std::cout << total << " lines chosen" << std::endl;
    
    vector<Vec2f> lines_chosen(total);
    place = 0;
    for (int i=0; (unsigned)i<lines_sparse.size(); i++) {
        if (yes_peaks[i]) {
            lines_chosen[place][0] = lines_sparse[i][0];
            lines_chosen[place][1] = lines_sparse[i][1];
            place++;
        }
    }
    
//     Mat disp_chosen = draw_lines(bw, lines_chosen);
//     namedWindow("chosen lines", CV_WINDOW_AUTOSIZE);
//     imshow("chosen lines", disp_chosen);
    
    // find parallel lines the right width apart
    // get indices
    int ii[lines_chosen.size()*lines_chosen.size()];
    int jj[lines_chosen.size()*lines_chosen.size()];
    place = 0;
    for (int i=0; (unsigned)i<lines_chosen.size(); i++) {
        float angle1 = lines_chosen[i][1]/(float)RADIANS_PER_DEGREE;
        float rho1 = lines_chosen[i][0];
        for (int j=0; j<i; j++) {
            float angle2 = lines_chosen[j][1]/(float)RADIANS_PER_DEGREE;
            float rho2 = lines_chosen[j][0];
            
            float tmpdiff = abs(angle1-angle2);
            float angd = min(tmpdiff, 180-tmpdiff);
            float dist = abs(rho1-rho2);
            
            if (dist<armw_max && dist>armw_min && angd<angd_tol) {
                ii[place] = i;
                jj[place] = j;
                place++;
            }
        }
    }
    total = place;
    std::cout << "num pairs: " << total << std::endl;

    // retrieve lines from indices
    vector<Vec2f> lines_chosen1(total);
    vector<Vec2f> lines_chosen2(total);
    for (int l=0; l<total; l++) {
        lines_chosen1[l][0] = lines_chosen[ii[l]][0];
        lines_chosen1[l][1] = lines_chosen[ii[l]][1];
        lines_chosen2[l][0] = lines_chosen[jj[l]][0];
        lines_chosen2[l][1] = lines_chosen[jj[l]][1];
    }
    
    // display all pairs
//     Mat disp;
//     vector<Vec2f> tmplines(2);
//     for (int i=0; (unsigned)i<lines_chosen1.size(); i++) {
//         namedWindow("pair " + ((char)i + '0'), CV_WINDOW_AUTOSIZE);
//         tmplines[0][0] = lines_chosen1[i][0];
//         tmplines[0][1] = lines_chosen1[i][1];
//         tmplines[1][0] = lines_chosen2[i][0];
//         tmplines[1][1] = lines_chosen2[i][1];
//         disp = draw_lines(bw, tmplines);
//         imshow("pair " + ((char)i + '0'), disp);
//     }

    // for each pair, find midline
    vector<Vec2f> midlines(lines_chosen1.size());
    for (int i=0; (unsigned)i<lines_chosen1.size(); i++) {
        midlines[i][0] = (lines_chosen1[i][0] + lines_chosen2[i][0])/2;
        midlines[i][1] = lines_chosen1[i][1];       // angles are supposed to be so close together that it won't matter.
    }
    Mat disp = draw_lines(bw, midlines);
    namedWindow("midlines", CV_WINDOW_AUTOSIZE);
    imshow("midlines", disp);
    
    // walk along midlines, find high gradients in right direction 
    // first get orientations
    Mat Dx(img.size(), CV_32FC1);
    Mat Dy(img.size(), CV_32FC1);
    Mat dMags(img.size(), CV_32FC1);
    int nconsider = 6;
    
    Mat kernel_x, kernel_y;
    kernel_x = dgauss(sig);
    kernel_y = dgauss(sig).t();
    flip(kernel_x, kernel_x, 1);    // convolution, not correlation, kernels
    flip(kernel_y, kernel_y, 0);    
    
    filter2D(imgf, Dx, Dx.depth(), kernel_x, Point(-1,-1), 0, BORDER_REPLICATE);
    filter2D(imgf, Dy, Dy.depth(), kernel_y, Point(-1,-1), 0, BORDER_REPLICATE);
    if (debug) {
        saveToFile<float>(Dx, "Dx_locateMid.txt");
        saveToFile<float>(Dy, "Dy_locateMid.txt");
    }
    
    magnitude(Dx, Dy, dMags);
    
    Point2f candidate_locs[midlines.size()*nconsider];
    float candMags[midlines.size()*nconsider];
    float candSims[midlines.size()*nconsider];
    
    cvtColor(bw, disp, CV_GRAY2RGB);    
    total=0;
    int candPlace = 0;
    for (int i=0; (unsigned)i<midlines.size(); i++) {
        float s = sin(midlines[i][1]);
        float c = cos(midlines[i][1]);
        std::cout << "cos=" << c << ", sin=" << s << std::endl;
        std::cout << "rho=" << midlines[i][0] << std::endl;
        
        float dx = s;
        float dy = -c;
        
        float stepx = dx/max(fabs(dx), fabs(dy));
        float stepy = dy/max(fabs(dx), fabs(dy));
//         std::cout << "max=" << max(fabs(dx), fabs(dy)) << ", fabs(dx)=" << fabs(dx) << ", fabs(dy)" << fabs(dy) << std::endl;
        std::cout << "stepx=" << stepx << ", stepy=" << stepy << std::endl;
        
        int sgn = 1;

        if (dx < 0 && left) {
            sgn = -1;
        } else if (dx > 0 && ~left) {
            sgn = -1;
        }
        
        std::cout << "dx=" << dx << ", dy=" << dy << std::endl;
        
        int nsteps = min(floor(fabs(nx/stepx)), floor(abs(ny/stepy)));
        float icept_y = midlines[i][0]/s;
        float icept_x = midlines[i][0]/c;
        std::cout << "icept_y " << icept_y << std::endl;
        std::cout << "icept_x " << icept_x << std::endl;
        
        float tx,ty;
        float gradMags[nsteps-2];
        float vals[nsteps-2];
        Mat tX(2, nsteps-2, CV_32FC1);
        Mat graddirs(2, nsteps-2, CV_32FC1);
        Mat gradDirSims(nsteps-2, 1, CV_32FC1);
        
        std::cout << "nsteps=" << nsteps << std::endl;
        for (int j=0; j<nsteps-2; j++) {// to be super-safe about boundaries
            if (left) {
                tx = dx*(j+1);
            } else {
                tx = nx + dx*(j+1);   
            }
            ty = dy*(j+1);

            if (icept_y >= 0) {
                ty = ty+icept_y;
            } else {
                tx = tx+icept_x;   
            }
            tX.at<float>(0,j) = tx;
            tX.at<float>(1,j) = ty;
//             std::cout << "tX=" << tx << " " << ty << std::endl;
            
            vals[j] = imgsm.at<float>(sround(ty), sround(tx));
            float gradx = Dx.at<float>(sround(ty),sround(tx));
            float grady = Dy.at<float>(sround(ty),sround(tx));
//             std::cout << "gradx=" << gradx << ", grady=" << grady << std::endl;
            float norm = sqrt(gradx*gradx + grady*grady);
            gradMags[j] = norm;
            if (norm > 0) {
                graddirs.at<float>(0,j) = gradx/norm;
                graddirs.at<float>(1,j) = grady/norm;
                gradDirSims.at<float>(j,0) = (dx*gradx + dy*grady)/norm;
            } else {
                graddirs.at<float>(0,j) = 0;
                graddirs.at<float>(1,j) = 0;
                gradDirSims.at<float>(j,0) = 0;
            }
        }
        
        if (debug) {
            saveArrayToFile<float>(gradMags, nsteps-2, "gradMags.txt");
            saveToFile<float>(graddirs, "graddirs.txt");
            saveToFile<float>(gradDirSims, "gradDirSims.txt");
            saveArrayToFile<float>(vals, nsteps-2, "vals.txt");
            saveToFile<float>(tX, "tX.txt");
        }
        
        int n_this_line=0;
        for (int j=0; j<nsteps-2; j++) {// to be super-safe about boundaries
            if ( (gradMags[j] > magthresh) 
                && (gradDirSims.at<float>(j,0) > simthresh)
                && ( j<=2 || abs(vals[j-3] - ex_gray) < gray_tol )) {
                    candidate_locs[candPlace].x = tX.at<float>(0,j);
                    candidate_locs[candPlace].y = tX.at<float>(1,j);
                    candMags[candPlace] = gradMags[j];
                    candSims[candPlace] = gradDirSims.at<float>(j,0);
                    n_this_line++;
                    candPlace++;
            }
            if (n_this_line >= nconsider) {
                break;   
            }
        }    
    } // for i from 0 to midlines.size()
    int numCands = candPlace;
    std::cout << "numCands " << numCands << std::endl;
    
    int midpt_ind;
    float min_score = 100000;
    for (int i=0; i<numCands; i++) {
        float diffx = candidate_locs[i].x - ex_loc.x;
        float diffy = candidate_locs[i].y - ex_loc.y;
        float dist = sqrt(diffx*diffx + diffy*diffy);
        float score = dist/candMags[i];
        
        if (score < min_score) {
            min_score = score;
            midpt_ind = i;
        }
    }

    if (numCands <= 0) {
        std::cout << "NO GOOD MATCHES FOUND" << std::endl;   
        return 1;
    }
    Point2f midpt(-1,-1);
    midpt.x = candidate_locs[midpt_ind].x;
    midpt.y = candidate_locs[midpt_ind].y;   
    std::cout << "MIDPOINT EQUALS (" << midpt.x << ", " << midpt.y << ")" << std::endl;
    
    cvtColor(imgsm256, disp, CV_GRAY2RGB);    
    for (int i=0; i<numCands; i++) {
        circle(disp, candidate_locs[i], 2, Scalar(0,0,255), 1);   
    }
    circle(disp, midpt, 1, Scalar(255, 255, 0), 1);
    namedWindow("midpt", CV_WINDOW_AUTOSIZE);
    imshow("midpt", disp);
    waitKey(0);
}

